File: src/Forge.Host/Bridge/BuildPayloadCodec.cs
namespace Forge.Host.Bridge;

public static class BuildPayloadCodec
{
    public static byte[] EncodeRequest(UyaBuildPatchRequestPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.ProjectRoot);
        writer.WriteString(value.CatalogRoot);
        writer.WriteString(value.CleanSourceIso);
        writer.WriteString(value.DevelopmentIso);
        writer.WriteString(value.SourceFingerprint);
        writer.WriteStrings(value.AcknowledgedWarnings);
        writer.WriteBoolean(value.ForceFullImage);
        writer.WriteStrings(value.IncludedLayers);
        return writer.ToArray();
    }

    public static UyaBuildPatchRequestPayload DecodeRequest(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new UyaBuildPatchRequestPayload(
            reader.ReadString(),
            reader.ReadString(),
            reader.ReadString(),
            reader.ReadString(),
            reader.ReadString(),
            reader.ReadStrings(),
            reader.ReadBoolean(),
            reader.ReadStrings());
        reader.Complete();
        return value;
    }

    public static byte[] EncodePlanRequest(UyaBuildPlanRequestPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.ProjectRoot);
        writer.WriteString(value.CatalogRoot);
        return writer.ToArray();
    }

    public static UyaBuildPlanRequestPayload DecodePlanRequest(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new UyaBuildPlanRequestPayload(reader.ReadString(), reader.ReadString());
        reader.Complete();
        return value;
    }

    public static byte[] EncodePlan(UyaBuildPlanPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteUInt32(checked((uint)value.Layers.Count));
        foreach (var layer in value.Layers)
        {
            writer.WriteString(layer.Layer);
            writer.WriteString(layer.State);
            writer.WriteBoolean(layer.CanDefer);
        }
        return writer.ToArray();
    }

    public static UyaBuildPlanPayload DecodePlan(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var count = reader.ReadUInt32();
        var layers = new UyaBuildLayerStatusPayload[count];
        for (var index = 0; index < layers.Length; index++)
            layers[index] = new(reader.ReadString(), reader.ReadString(), reader.ReadBoolean());
        reader.Complete();
        return new(layers);
    }

    public static byte[] EncodeProgress(UyaBuildPatchProgressPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.Phase);
        writer.WriteUInt64(value.Completed);
        writer.WriteUInt64(value.Total);
        writer.WriteString(value.Message);
        return writer.ToArray();
    }

    public static UyaBuildPatchProgressPayload DecodeProgress(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new UyaBuildPatchProgressPayload(
            reader.ReadString(), reader.ReadUInt64(), reader.ReadUInt64(), reader.ReadString());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeResult(UyaBuildPatchResultPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteBoolean(value.Succeeded);
        writer.WriteBoolean(value.RequiresWarningAcknowledgement);
        writer.WriteStrings(value.WarningCodes);
        writer.WriteStrings(value.Diagnostics);
        writer.WriteString(value.Message);
        writer.WriteString(value.NextAction);
        writer.WriteString(value.DevelopmentIsoPath);
        writer.WriteString(value.PatchMode);
        writer.WriteString(value.OutputLevelWadSha256);
        writer.WriteUInt32(value.BakedLayerCount);
        writer.WriteBoolean(value.BakeWasCurrent);
        return writer.ToArray();
    }

    public static UyaBuildPatchResultPayload DecodeResult(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new UyaBuildPatchResultPayload(
            reader.ReadBoolean(),
            reader.ReadBoolean(),
            reader.ReadStrings(),
            reader.ReadStrings(),
            reader.ReadString(),
            reader.ReadString(),
            reader.ReadString(),
            reader.ReadString(),
            reader.ReadString(),
            reader.ReadUInt32(),
            reader.ReadBoolean());
        reader.Complete();
        return value;
    }
}

File: src/main/bridge/BuildPayloadCodec.ts
import { Buffer } from 'node:buffer';

import type { UyaBuildPatchRequest, UyaBuildPlanRequest } from '../../types/BridgePayloads.js';
import type { BuildLayerId, BuildPatchProgress, BuildPatchResult, BuildPlan } from '../../types/ForgeApi.js';
import { PayloadReader, PayloadWriter } from './PayloadIO.js';

export function encodeBuildPatchRequest(value: UyaBuildPatchRequest): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.projectRoot);
  writer.writeString(value.catalogRoot);
  writer.writeString(value.cleanSourceIso);
  writer.writeString(value.developmentIso);
  writer.writeString(value.sourceFingerprint);
  writer.writeStrings(value.acknowledgedWarnings);
  writer.writeBoolean(value.forceFullImage);
  writer.writeStrings(value.includedLayers);
  return writer.toBuffer();
}

export function encodeBuildPlanRequest(value: UyaBuildPlanRequest): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.projectRoot);
  writer.writeString(value.catalogRoot);
  return writer.toBuffer();
}

export function decodeBuildPlan(payload: Uint8Array): BuildPlan {
  const reader = new PayloadReader(payload);
  const count = reader.readUInt32();
  const layers = Array.from({ length: count }, () => ({
    layer: reader.readString() as BuildLayerId,
    state: reader.readString() as BuildPlan['layers'][number]['state'],
    canDefer: reader.readBoolean(),
  }));
  reader.complete();
  return { layers };
}

export function decodeBuildPatchProgress(payload: Uint8Array): BuildPatchProgress {
  const reader = new PayloadReader(payload);
  const value = {
    phase: reader.readString(),
    completed: reader.readUInt64(),
    total: reader.readUInt64(),
    message: reader.readString(),
  };
  reader.complete();
  return value;
}

export function decodeBuildPatchResult(payload: Uint8Array): BuildPatchResult {
  const reader = new PayloadReader(payload);
  const succeeded = reader.readBoolean();
  const requiresWarningAcknowledgement = reader.readBoolean();
  const warningCodes = reader.readStrings();
  const diagnostics = reader.readStrings();
  const message = reader.readString();
  const nextAction = reader.readString();
  const developmentIsoPath = reader.readString() || undefined;
  const patchModeValue = reader.readString();
  const outputLevelWadSha256 = reader.readString() || undefined;
  const bakedLayerCount = reader.readUInt32();
  const bakeWasCurrent = reader.readBoolean();
  reader.complete();
  const patchMode = patchModeValue === 'InPlace' || patchModeValue === 'FullImageReplacement'
    ? patchModeValue : undefined;
  return {
    succeeded,
    requiresWarningAcknowledgement,
    warningCodes,
    diagnostics,
    message,
    nextAction,
    developmentIsoPath,
    patchMode,
    outputLevelWadSha256,
    bakedLayerCount,
    bakeWasCurrent,
  };
}
